service/matching/taskQueueManager.go

```diff
@@ -426,14 +455,65 @@ func (c *taskQueueManagerImpl) DispatchQueryTask(
 	return c.matcher.OfferQuery(ctx, task)
 }
 
+// GetVersioningData returns the versioning data for the task queue if any. If this task queue is a sub-partition and
+// has no cached data, it will explicitly attempt a fetch from the root partition.
 func (c *taskQueueManagerImpl) GetVersioningData(ctx context.Context) (*persistencespb.VersioningData, error) {
-	return c.db.GetVersioningData(ctx)
+	vd, err := c.db.GetVersioningData(ctx)
+	if errors.Is(err, errVersioningDataNotPresentOnPartition) {
+		// If this is a non-root-partition with no versioning data, this call is indicating we might expect to find
+		// some. Since we may not have started the poller, we want to explicitly attempt to fetch now, because
+		// it's possible some was added to the root partition, but it failed to invalidate this partition.
+		return c.fetchMetadataFromRootPartition(ctx)
+	}
+	return vd, err
 }
 
 func (c *taskQueueManagerImpl) MutateVersioningData(ctx context.Context, mutator func(*persistencespb.VersioningData) error) error {
 	err := c.db.MutateVersioningData(ctx, mutator)
 	c.signalIfFatal(err)
-	return err
+	if err != nil {
+		return err
+	}
+	// If this is the root partition, notify partitions that they should fetch changed data from us.
+	if c.taskQueueID.IsRoot() {
+		numParts := util.Max(c.config.NumReadPartitions(), c.config.NumWritePartitions())
+		wg := &sync.WaitGroup{}
+		for i := 1; i < numParts; i++ {
+			wg.Add(1)
+			go func(i int) {
+				tq := c.taskQueueID.mkName(i)
+				_, err := c.matchingClient.InvalidateTaskQueueMetadata(ctx,
+					&matchingservice.InvalidateTaskQueueMetadataRequest{
+						NamespaceId:    c.taskQueueID.namespaceID.String(),
+						TaskQueue:      tq,
+						VersioningData: true,
```

Comment:
why ask leaf partition to make additional poll, can we pass in the new data here directly?

Reply:
Mostly just a matter of always having the data come from one place which is clean and simple. We can change it to work that way and save a round trip though.

Reply:
Changed.